### PR TITLE
Added broker implementation

### PR DESCRIPTION
**Pull request by gavraz**

### CODE REVIEW

File: hare/hare_shared.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+import (
+	"github.com/spacemeshos/go-spacemesh/p2p/service"
+	"hash/fnv"
+	"time"
+)
+
+const ProtoName = "HARE_PROTOCOL"
+const RoundDuration = time.Second * time.Duration(15)
+const (
+	Status   = 0 // round 0
```
Comment (almogdepaz) [inline hare/hare_shared.go]: i think its better if we define this enum in either the proto file or here
doing it twice may cause issues in the future

File: hare/broker.go [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+			err := proto.Unmarshal(msg.Data(), hareMsg)
+			if err != nil {
+				log.Error("Could not unmarshal message: ", err)
+			}
+
+			h := fnv.New32()
+			h.Write(hareMsg.Message.Layer)
+			id := h.Sum32()
+
+			broker.outbox[id] <- hareMsg
```
Comment (antonlerner) [inline hare/broker.go]: please sync reading from map with read mutex

File: hare/broker.go [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+// Inbox returns the message channel associated with the given layer
+func (broker *Broker) Inbox(iden Identifiable) chan *pb.HareMessage {
+	broker.mutex.Lock()
+	defer broker.mutex.Unlock()
+
+	var id = iden.Id()
+	if _, exist := broker.outbox[id]; exist {
+		panic("Inbox called more than once per layer")
+	}
+
```
Comment (kenze1) [inline hare/broker.go]: sync writing to map with write mutex

File: hare/broker.go [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+		case <-broker.StopChannel():
+			return
+		}
+	}
+}
+
+// Inbox returns the message channel associated with the given layer
+func (broker *Broker) Inbox(iden Identifiable) chan *pb.HareMessage {
+	broker.mutex.Lock()
```
Comment (antonlerner) [inline hare/broker.go]: you can lock this later

File: hare/broker.go [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+	defer broker.mutex.Unlock()
+
+	var id = iden.Id()
+	if _, exist := broker.outbox[id]; exist {
+		panic("Inbox called more than once per layer")
```
Comment (antonlerner) [inline hare/broker.go]: is it worth panicing here?

File: hare/broker.go [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+			broker.outbox[id] <- hareMsg
+
+		case <-broker.StopChannel():
+			return
+		}
+	}
+}
+
+// Inbox returns the message channel associated with the given layer
+func (broker *Broker) Inbox(iden Identifiable) chan *pb.HareMessage {
```
Comment (antonlerner) [inline hare/broker.go]: Rename to CreateInbox

File: hare/broker.go (hunk 2 above)
Comment (gavraz) [inline hare/broker.go]: Verify exist.

File: hare/broker.go [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+	channel chan struct{} // listeners listen to this channel
+}
+
+func NewStopper(listenersCount uint8) *Stopper {
+	return &Stopper{listenersCount, make(chan struct{})}
+}
+
+// Stops all listening instances (should be called only once)
+func (stopper *Stopper) Stop() {
+	for i := uint8(0); i < stopper.count; i++ {
```
Comment (antonlerner) [inline hare/broker.go]: you can use the close(ch) func to signal all listeners on this channel that it has closed

File: hare/broker.go [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+	for {
+		select {
+		case msg := <-broker.inbox:
+			hareMsg := &pb.HareMessage{}
+			err := proto.Unmarshal(msg.Data(), hareMsg)
+			if err != nil {
+				log.Error("Could not unmarshal message: ", err)
+			}
+
+			h := fnv.New32()
```
Comment (antonlerner) [inline hare/broker.go]: add comment describing logic

File: hare/broker.go [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+	*Stopper
+	network NetworkService
+	inbox   chan service.Message
+	outbox  map[uint32]chan *pb.HareMessage
+	mutex   sync.Mutex
+}
+
+func NewBroker(networkService NetworkService) *Broker {
+	p := new(Broker)
+	p.Stopper = NewStopper(1)
```
Comment (kenze1) [inline hare/broker.go]: receive magic number '1' as parameter to constructor

File: hare/broker.go [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+}
+
+// StopChannel returns the channel channel to wait on
+func (stopper *Stopper) StopChannel() chan struct{} {
+	return stopper.channel
+}
+
+// Increment the number of listening instances
+// Should not be called after Stop()
+func (stopper *Stopper) Increment() {
```
Comment (antonlerner) [inline hare/broker.go]: remove this func or comment that it is for DEBUG purposes only

File: hare/broker.go (hunk 5 above)
Comment (gavraz) [reply hare/broker.go]: Currently calling twice to this func is a wrong usage.

File: hare/broker.go [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+func (stopper *Stopper) StopChannel() chan struct{} {
+	return stopper.channel
+}
+
+// Broker is responsible for dispatching hare messages to the matching layer listener
+type Broker struct {
+	Stopper
+	network NetworkService
+	inbox   chan service.Message
+	outbox  map[uint32]chan *pb.HareMessage
```
Comment (gavraz) [inline hare/broker.go:37]: I should probably add a method to remove chan from the map once a consensus instance terminates (if not, the map will only grow). I will address it in the other part of the initial implementation task.

File: hare/broker.go [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+
+	if _, exist := broker.outbox[id]; exist {
+		panic("CreateInbox called more than once per layer")
+	}
+
+	broker.mutex.Lock()
+	broker.outbox[id] = make(chan *pb.HareMessage, InboxCapacity) // create new channel
+	broker.mutex.Unlock()
+
+	return broker.outbox[id]
```
Comment (antonlerner) [inline hare/broker.go]: lock read from shared map

File: hare/broker.go [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+	"github.com/spacemeshos/go-spacemesh/hare/pb"
+	"github.com/spacemeshos/go-spacemesh/log"
+	"github.com/spacemeshos/go-spacemesh/p2p/service"
+	"sync"
+)
+
+const InboxCapacity = 100
+
+// Stopper is used to add stoppability to an object
+type Stopper struct {
```
Comment (y0sher) [inline hare/broker.go]: Isn't that an over kill ? why not just have a channel and close it ?

File: hare/broker.go [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+			return
+		}
+	}
+}
+
+// CreateInbox creates and returns the message channel associated with the given layer
+func (broker *Broker) CreateInbox(iden Identifiable) chan *pb.HareMessage {
+	var id = iden.Id()
+
+	if _, exist := broker.outbox[id]; exist {
```
Comment (antonlerner) [inline hare/broker.go:94]: add a lock here as well

File: hare/broker.go [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+
+// Dispatch incoming messages to the matching layer instance
+func (broker *Broker) dispatcher() {
+	for {
+		select {
+		case msg := <-broker.inbox:
+			hareMsg := &pb.HareMessage{}
+			err := proto.Unmarshal(msg.Data(), hareMsg)
+			if err != nil {
+				log.Error("Could not unmarshal message: ", err)
```
Comment (antonlerner) [inline hare/broker.go:70]: should you process with flow if message parsing failed?

File: hare/broker.go [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+	p := new(Broker)
+	p.Stopper = NewStopper()
+	p.network = networkService
+	p.outbox = make(map[uint32]chan *pb.HareMessage)
+
+	return p
+}
+
+// Start listening to protocol messages and dispatch messages
+func (broker *Broker) Start() {
```
Comment (antonlerner) [inline hare/broker.go:51]: what happens if start is called twice?

File: hare/broker.go [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+type Stopper struct {
+	channel chan struct{} // stoppable go routines listen to this channel
+}
+
+func NewStopper() Stopper {
+	return Stopper{make(chan struct{})}
+}
+
+// Stops all listening instances (should be called only once)
+func (stopper *Stopper) Stop() {
```
Comment (y0sher) [inline hare/broker.go]: Close is the convention

File: hare/broker.go (hunk 15 above)
Comment (gavraz) [reply hare/broker.go:70]: Nope. Added continue.

File: hare/broker.go [hunk 18]
```diff
@@ -0,0 +1,105 @@
...
+				continue
+			}
+
+			layerId := NewLayerId(hareMsg.Message.Layer)
+
+			broker.mutex.RLock()
+			if c, exist := broker.outbox[layerId.Id()]; exist {
+				c <- hareMsg
+			}
+			broker.mutex.RUnlock()
```
Comment (antonlerner) [inline hare/broker.go]: can channel c be blocked causing the mutex not to be released?

File: hare/broker.go [hunk 19]
```diff
@@ -0,0 +1,105 @@
...
+		panic("CreateInbox called more than once per layer")
+	}
+	broker.mutex.RUnlock()
+
+	broker.mutex.Lock()
+	broker.outbox[id] = make(chan *pb.HareMessage, InboxCapacity) // create new channel
+	broker.mutex.Unlock()
+
+	broker.mutex.RLock()
+	defer broker.mutex.RUnlock()
```
Comment (antonlerner) [inline hare/broker.go]: why do it like that if you're just returning the value? you can alternatively just return the new channel you just created, why read it from the map again?

File: hare/hare_shared.go [hunk 20]
```diff
@@ -0,0 +1,43 @@
+package hare
```
Comment (antonlerner) [inline hare/hare_shared.go:1]: why did you call it hare_shared? what is the purpose of splitting this code from the hare protocol?

**Review by antonlerner — APPROVED** [review]

**Pull request merged**